src/cpp/core/system/PosixGroup.cpp

```diff
@@ -121,13 +117,12 @@ Error userGroups(const std::string& userName, std::vector<Group>* pGroups)
    typedef int GIDTYPE;
 #endif
 
-   boost::shared_ptr<GIDTYPE> pGids(new GIDTYPE[100]);
-   while (!getgrouplist(userName.c_str(), user.getGroupId(), pGids.get(), &numGroups))
+   // get the groups for the user - we start with 100 groups which should be enough for most cases
+   // if it is not, resize the buffer with the correct amount of groups and try again
+   int numGroups = 100;
+   boost::shared_ptr<GIDTYPE> pGids(new GIDTYPE[numGroups]);
+   while (getgrouplist(userName.c_str(), user.getGroupId(), pGids.get(), &numGroups) == -1)
```

Comment:
FWIW I think we can remove the while loop here and just have two explicit calls to `getgrouplist()`. If I'm reading the documentation correctly, if the function returns -1, then the real required number of groups will be written to `&numGroups` and we can just make one more call with that value. (It also gives us a chance to sanity-check the second call to `getgrouplist()`, which should in theory never fail)

Reply:
Yes, I thought of doing that as well but figured it was ok to leave as is. I'll update with this change.

Reply:
I am torn though because this introduces the possibility of the following sequence:

1. Initial call to `getgrouplist`
2. User is added to another group via some other mechanism
3. Second call to `getgrouplist` still fails, since the count parameter is wrong

This is not likely but it is one of the reasons I decided to keep the `while` loop - what do you think?

Reply:
Ah, that's a very good point. I think that's motivation enough to keep things as-is.